common/notification/index.ts

```diff
@@ -0,0 +1,216 @@
+        }
+
+        const notification = await getNotification(reminder.notificationID);
+        const user = await getUser(reminder.userId);
+        await deliverNotification(reminder, notification, user, reminder.context as NotificationContext);
+
+        // TODO: What about intervals?
+    }
+
+    info(`Sent ${remindersToSend.length} reminders in ${Date.now() - start}ms`);
+
+}
+
+// TODO: function for user preferences "categories"
+// TODO: Check queue state, find pending emails and ones with errors, report to Admins, resend / cleanup utilities
+
+/* --------------------------- Concrete Notification "Queue" ----------------------------------- */
+
+/* Creates an entry in the concrete_notifications table, to track the notification */
+async function createConcreteNotification(notification: Notification, user: Person, context: NotificationContext): Promise<ConcreteNotification> {
+    if (context.uniqueId) {
+        const existingNotification = await prisma.concrete_notification.findFirst({
+            where: {
+                notificationID: notification.id,
+                userId: getUserId(user),
+                contextID: context.uniqueId
+            }
+        });
+
+        if (existingNotification) {
+            throw new Error(`Duplicate Notification(${notification.id}) for User(${user.id}) with Context(${context.uniqueId})`);
+        }
+    }
+
+    // First of all we commit the notification to the database, which allows us to recover if the backend crashes
+    const concreteNotification = await prisma.concrete_notification.create({
+        data: {
+            // the unique id is automatically created by the database
+            notificationID: notification.id,
+            state: ConcreteNotificationState.PENDING,
+            userId: getUserId(user),
+            sentAt: new Date(),
+            contextID: context.uniqueId,
+            context
+        }
+    });
+
+    debug(`Notification.createConcreteNotification succeeded for ConcreteNotification(${concreteNotification.id})`);
+
+    return concreteNotification;
+}
+
+async function deliverNotification(concreteNotification: ConcreteNotification, notification: Notification, user: Person, notificationContext: NotificationContext): Promise<void> {
+    debug(`Sending ConcreteNotification(${concreteNotification.id}) of Notification(${notification.id}) to User(${user.id})`);
+
+    const context: Context = {
+        ...notificationContext,
+        user: { ...user, fullName: getFullName(user) }
+    };
+
+    try {
+        const channel = channels.find(it => it.canSend(notification.id));
+        if (!channel) {
+            throw new Error(`No fitting channel found for Notification(${notification.id})`);
+        }
+
+        // TODO: Check if user silenced this notification
+
+        await channel.send(notification.id, user, context, concreteNotification.id);
```

Comment:
I considered this and there was a version were we had dedicated IDs per Channel. After some consideration I removed it again, as currently every Notification is a Mailjet Notification, and AFAIK the SMS Channel will only be used as an alternative Channel for the same notifications. Thus we currently have a 1:1 Mapping between Notifications and Mailjet Templates. As such the "Template ID" is a unique ID for the Notification and a good candidate for a primary key. 

I'd prefer to leave it as is for now and reconsider this once we have more than the Mailjet channel, but I'm also open to change this as part of this PR.

Reply:
Yeah, we originally had this, then removed it because the TemplateID was already a unique id of the Notification, though you're right that this creates a tight coupling between Template and the Notification which might get in the way in the future. Adding this would also open up the possibility for Notifications that cannot be delivered via the Mailjet Channel (TemplateID: NULL). I'm gonna adapt this.